### DISCUSSION TITLE
Font selection order dependence

### DISCUSSION THREAD

**Discussion by KillTheMule** [typst/typst#3893] [General]
I'm using typst as a library, and I've recently changed how my binary includes fonts (I don't use any system or dynamic fonts, only those included in the binary). I'm still using the exact same fonts and the exact same typst documents (however, the internal order of the fonts surely has changed).

<details>
<summary>Quick overview</summary>
Before, I included a font collection I made, and the binary would extract the fonts in a loop, create a `Vec<Font>` from this and then create a `FontBook` from that. Both were elements of a struct implementing `World` in the obvious way for fonts.

Now, I include each font separately, and run

```rust
          let buffer = typst::foundations::Bytes::from_static(data);
          for font in Font::iter(buffer) {
              book.push(font.info().clone());
              fonts.push(font);
          }
```
over its `data` to create the same as above.

Note that the order of the fonts in the `Vec<Font>` and consequently the `FontBook` has most probably changed (I can't be sure because I don't fully understand font collections).
</details>

After the change, some spacing changed in my documents. It's not a real problem, but I'd like to understand where it's coming from.

One thing I noted is that in particular some empty lines (produced via `linebreak()`) did change their height, so one working theory I have is that the choice of font for parts of the document has been changed. Note however that the default test font should be an exact match.

I have two questions:

1. How could I investigate this? I don't know how to see which fonts exactly have been chosen for what part of the document (and maybe which scaling was applied or such).
2. Should the internal order of the fonts matter for that? I'd have guessed typst searches all available fonts for matches and takes what it considers best, which should not depend on order...

Thanks for any pointers, if details are of interest I'd provide them for sure :)

**Comment by Myriad-Dreamin** [comment]
1. You can see the text elements for font usage in the `typst::model::Document`, https://github.com/Myriad-Dreamin/tinymist/blob/da70c0d31c30e1a6fb69141803a4f244efe3d64f/crates/tinymist-query/src/document_metrics.rs#L127-L156
3. The order of you adding fonts to the `typst::text::FontBook` does matter. Since if there are multiple fonts meeting conditions during selecting fonts, the first one is used.

**Reply by KillTheMule** [reply]
Thanks for your answers, I'll see how far I can get analyzing font usage like you linked. The second part makes sense, but I feel like it shouldn't apply here, because I have so few fonts... let's see what turns up :)

(e) How the heck to I mark your post as the answer? Sorry, can't find the right button ;(

**Reply by KillTheMule** [reply]
Small update: Your code helped me identify that indeed the font choice was different, and further investingation revealed that fonts that I though "the same" actually were "not really the same". Restoring the old font files fixed my small issue, and I understood how it happened. Thanks again!
